Translate subset of Wallet API to Servant

jonathanknowles opened the pull request:
Translate a subset of the Wallet API to Servant. (See Issue #53)

# Included in this PR

This PR:

* translates a small number of endpoints within the `wallets` namespace.
* translates all types associated with the above endpoints.
* provides JSON encodings that (should) match the Swagger type definitions.
* provides round-trip validation for all types.
* provides round-trip tests for JSON serialization and deserialization.

# Not included in this PR

This PR:

* doesn't attempt to encode the entire API. Instead, a small subset is encoded.
* doesn't attempt to automatically verify that the Servant definition matches the Swagger definition.

# Design choices ##

This PR makes some design choices, which should be reviewed.

## 1. Place each type in a separate module

Reasons:    

* Better encapulation and type safety.
  Example: `WalletName` cannot be constructed without a valid name.
* GHC can parallelize builds of types within separate modules.

## 2. Favour simple generic JSON encoding

Manually-written JSON code can be tricky to debug.

This PR makes the choice of using generically-derived `ToJSON` and `FromJSON` instances wherever possible.

Whenever there's a non-trivial mapping between a Haskell type and a desired encoding, we encode the desired structure in a new Haskell type that's closer to the desired encoding, and write a pair of conversion functions that can easily be checked by the compiler.

For example: `WalletState` provides an idiomatic Haskell encoding of a wallet's restoration state, but we also provide `UnvalidatedWalletState` as a type that is more trivial to encode as JSON. We provide `validate` and `unvalidate` to convert between these two types.

jonathanknowles committed:
Add partial Servant specification of wallet API.

jonathanknowles committed:
Represent `Percentage` with a rich JSON object rather than a naked value.

jonathanknowles committed:
Simplify `Arbitrary` instance for `WalletName`.

jonathanknowles committed:
Move `uuidFromWords` closer to its usage point.

jonathanknowles committed:
Add more round-trip JSON serialization tests.

jonathanknowles committed:
Add `GetWallet` to the API.

jonathanknowles committed:
Add `DeleteWallet` to API.

jonathanknowles committed:
Remove unnecessary dependencies.

jonathanknowles committed:
Change imports to qualified.

jonathanknowles committed:
Remove unnecessary `Functions` module.

jonathanknowles committed:
Remove `Unknown` value of `CurrencyUnit`.

jonathanknowles committed:
Make the names of functions in `Api.V2.JSON` more self-describing.

jonathanknowles committed:
Fix order of imports.

jonathanknowles committed:
Satisfy import rules in module `Api.V2.Lenses`.

jonathanknowles committed:
Export all values of `CurrencyUnit` (there is only one).

jonathanknowles commented on src/Cardano/Wallet/Api/V2/Types/CurrencyUnit.hs:
```diff
@@ -0,0 +1,22 @@
+    parseJSON (String "lovelace") = pure Lovelace
+    parseJSON _ = fail "Invalid currency unit"
+instance ToJSON CurrencyUnit where
+    toJSON Lovelace = String "lovelace"
```
Here we encode the "lovelace" string manually, as the `ObjectWithSingleField` option for `sumEncoding` does not work for types with only a single constructor.

KtorZ replied:
```hs
λ> data CurrencyUnit = Lovelace deriving (Generic, Show)
λ> let opts = defaultOptions { tagSingleConstructors = True, constructorTagModifier = map toLower }
λ> genericToJSON opts Lovelace
String "lovelace"
```

:wink: 

The `SumEncoding` tells `aeson` how different branches of a sum type should be encoded, though, how each constructor for those branches are encoded is specified by `constructorTagModifier`. Yet, for single constructor, we need to have the extra `tagSingleConstructors` enabled otherwise, `aeson` encode them to an empty `Array` (god knows why).

jonathanknowles replied:
Thanks! This magic combination seems to work for both sum types and types with just a single constructor. I'll use this as the definition of `defaultSumTypeOptions`.

KtorZ commented on src/Cardano/Wallet/Api/V2/Lenses.hs:
```diff
@@ -0,0 +1,25 @@
+makeFieldsNoPrefix ''T.WalletDelegation
+makeFieldsNoPrefix ''T.WalletId
+makeFieldsNoPrefix ''T.WalletPassphraseInfo
+makeFieldsNoPrefix ''T.WalletState
```
I'll suggest using [`generic-lens`](https://hackage.haskell.org/package/generic-lens) for lenses and, before introducing this, have a proper discussion with the team about whether we want to use lenses and, what features from the lenses we allow. 

We had quite a hard time dealing with the old code base where everything with a Zoomable of a Traversal after going through 3 different Prisms. All of that just to modify a field in the DB :thinking: ...

jonathanknowles replied:
Fair enough! I've removed this module for the moment, and have removed the dependency on `lens`.

KtorZ commented on src/Cardano/Wallet/Api/V2/Types/WalletAddressPoolGap.hs:
```diff
@@ -0,0 +1,52 @@
+    | j > maxBound = Left WalletAddressPoolGapTooLargeError
+    | otherwise = pure j
+  where
+    j = WalletAddressPoolGap $ fromIntegral i
```
For this and the `Bounded` instance, we _should_ rather rely on the underling `AddressPoolGap` type to avoid duplicating logic and for them becoming out of sync rather soon.

jonathanknowles replied:
Understood! I've based this type on the preexisting `AddressPoolGap` type (which came into existence after I began this PR).

I'm not sure if I'm happy with the result (yet). If we have to wrap several types, then perhaps there is a more elegant method.

KtorZ commented on src/Cardano/Wallet/Api/V2/Types/WalletId.hs:
```diff
@@ -0,0 +1,14 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
When going for `GeneralizedNewtypeDeriving` or `DeriveAnyClass`, I'd advise to always add `DerivingStrategies` and then, specify which strategy is used to derive things (`stock`, `newtype` or `any`):

e.g.

```hs
newtype WalletId = WalletId
    { _uuid :: UUID }
    deriving stock (Eq, Show)
    deriving newtype (FromJSON, ToJSON)
```

We've already limited the impact of this by requiring each module to specify their own extensions, but this will also cover our back when potentially non-desirable things are done through automatic deriving.

jonathanknowles replied:
Thanks, I'll change this (and any others) to match your suggestion.

KtorZ commented on cardano-wallet.cabal:
```diff
@@ -66,6 +70,22 @@ library
       Cardano.Wallet
       Cardano.Wallet.AddressDerivation
       Cardano.Wallet.AddressDiscovery
+      Cardano.Wallet.Api.V2
```
Let's drop the `V2` prefix from the module's name. This has already caused us pain with the `V1` api. Beside, it's technically the V4 of the API if we are really honest :confused:  
Let's don't have any prefix in our module hierarchy.

jonathanknowles replied:
Understood! Will drop the prefix.

jonathanknowles committed:
Apply hlint suggestions.

KtorZ commented on cardano-wallet.cabal:
```diff
@@ -66,6 +70,22 @@ library
+      Cardano.Wallet.Api.V2.Types.WalletName
+      Cardano.Wallet.Api.V2.Types.WalletPassphraseInfo
+      Cardano.Wallet.Api.V2.Types.WalletState
+      Cardano.Wallet.Api.V2.Types.WalletStateStatus
```
Hehe. Many, Tiny, Modules. May we group that into one module called `Types`, unless there's a good reason to split something in a separate one. Again, we've done that, aaand, having many modules that live within the same "realm" and are at the same layers doesn't work well. Modules get rapidly inconsistent and, it's quite painful to maintain and have to jump between all of them when making a change. Having one module with all the API types is fine.

jonathanknowles committed:
Apply weeder suggestions.

KtorZ commented on test/unit/Cardano/Wallet/Api/V2/Arbitraries.hs:
```diff
@@ -0,0 +1,97 @@
+
+instance Arbitrary WalletId where
+    arbitrary = WalletId . uuidFromWords <$> arbitrary
+    shrink (WalletId u) = WalletId . uuidFromWords <$> shrink (UUID.toWords u)
```
What' the rational behind shrinking a UUID :thinking: ?

jonathanknowles replied:
I can't think of any. I'll remove it.

KtorZ commented on test/unit/Cardano/Wallet/Api/V2/Arbitraries.hs:
```diff
@@ -0,0 +1,97 @@
+        nameLength <- choose (walletNameMinLength, walletNameMaxLength)
+        fromRight (error "Unable to create arbitrary WalletName")
+            . mkWalletName
+            . Text.pack <$> replicateM nameLength (choose ('a', 'z'))
```
Here, a shrinker that rapidly yields the shortest possible name would be useful :sweat_smile:

jonathanknowles replied:
Done!

jonathanknowles committed:
Simplify JSON encoding of `CurrencyUnit`.

jonathanknowles committed:
Rename `Api.V2` to `Api`.

jonathanknowles committed:
Add test to check that we can perform basic JSON deserialization of an
example wallet.

jonathanknowles committed:
Use `DerivingStrategies` in definition of `WalletId`.

jonathanknowles committed:
Use `DerivingStrategies` in definition of `WalletAddressPoolGap`.

jonathanknowles committed:
Use `DerivingStrategies` in definition of `WalletName`.

jonathanknowles committed:
Don't shrink `WalletId` in definition of `Arbitrary`.

jonathanknowles committed:
Provide an `Arbitrary.shrink` implementation for `WalletName`.

jonathanknowles committed:
Expand the range of arbitrary wallet names to include all printable
Unicode characters.

jonathanknowles committed:
Remove weed from `Cardano.Wallet.Api.JsonSpec`.

jonathanknowles committed:
Remove `Cardano.Wallet.Api.Lenses`.

jonathanknowles committed:
Remove lens dependency.

jonathanknowles committed:
Reuse existing `AddressPoolGap` type in wallet API.

rvl commented on src/Cardano/Wallet/Api.hs:
```diff
@@ -0,0 +1,31 @@
+api = Proxy
+
+type DeleteWallet = "wallets"
+    :> "deleteWallet"
```
Looking at the `swagger.yaml`, I don't think `deleteWallet`, `getWallet`, `listWallets`, etc, are in the request paths for these API endpoints.

jonathanknowles replied:
Well spotted! I've removed these.

rvl commented on src/Cardano/Wallet/Api/Types/Percentage.hs:
```diff
@@ -0,0 +1,70 @@
+
+data UnvalidatedPercentage = UnvalidatedPercentage
+    { _quantity :: Int
+    , _unit :: Text
```
I know it matches the `swagger.yaml` specification, but this `unit: "percent"` thing is hokey. There are no other units than percent. I reckon the property name should just be changed. E.g. `progress` → `progressPercent`.

jonathanknowles replied:
You're right that there are no units other than `percent`.

I used the `UnvalidatedPercentage` type in order to make JSON encoding easier, because it's very close to the JSON representation. As part of validation during deserialization, we transform an `UnvalidatedPercentage` to a `Percentage`, which is the idiomatic Haskell representation.

Notice that we only export `Percentage` from the module, and not `UnvalidatedPercentage` (values of which only exist transiently).

rvl commented on src/Cardano/Wallet/Api/Types/WalletDelegation.hs:
```diff
@@ -0,0 +1,23 @@
+
+newtype WalletDelegation = WalletDelegation
+    { _status :: T.WalletDelegationStatus
+    } deriving (Eq, Generic, Show)
```
Is this going to have the `target` property, or will we do that later?

jonathanknowles replied:
Well spotted. I'll add this in.

rvl commented on src/Cardano/Wallet/Api/Types/WalletState.hs:
```diff
@@ -0,0 +1,62 @@
+instance ToJSON UnvalidatedWalletState where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+validate :: UnvalidatedWalletState -> Either WalletStateError WalletState
```
Is this validation needed? I think the user can POST or PUT a wallet, but they can't update the `state` property.

jonathanknowles replied:
We might write clients in Haskell (or write tests of round-trip functionality). In such a case, this functionality would used during deserialization on the client end.

Of course, if this isn't needed, then we could also remove the `FromJSON` instance.

rvl replied:
OK, yes I didn't consider clients yet.

rvl commented on test/unit/Cardano/Wallet/Api/JsonSpec.hs:
```diff
@@ -0,0 +1,87 @@
+spec = do
+    describe "can perform basic JSON deserialization" $
+        it "Wallet" $
+            (eitherDecode exampleWallet :: Either String Wallet)
```
I don't think this matches how the API will be used. The test should be the other way around.
```
exampleWalletBytes <- B8.readFile "example-wallet.json"
encode exampleWallet `shouldBe` exampleWalletBytes
```

The wallet objects that users put and post have less fields, and will want to have decoding tests.

KtorZ replied:
For testing, we'll need:

- Golden tests for every types representation in JSON (as @rvl just described, this will ensure that we are at least aware of breaking changes). Ideally, we could _generate_ golden tests from arbitrary instances to have several ones for each Api types.
- JSON roundtrips 
- HttpApiData roundtrips (when it make sense)
- Swagger schema comparison with the specification

(not necessarily all in this PR though!)

jonathanknowles replied:
It's correct that this wouldn't normally be used by a user. However, having this test helps us to write Haskell clients against this API (which may occur as part of a future test suite).

jonathanknowles replied:
> The wallet objects that users put and post have less fields, and will want to have decoding tests.

Agreed! We can include these in the next PR.

rvl reviewed (changes requested):
Looks pretty good. I have a few questions and one comment.

jonathanknowles committed:
Remove Swagger operation IDs from Servant endpoint paths.

KtorZ commented on src/Cardano/Wallet/Api.hs:
```diff
@@ -0,0 +1,28 @@
+
+type DeleteWallet = "wallets"
+    :> Capture "walletId" T.WalletId
+    :> Delete '[] ()
```
I think we still want the content type to get the headers set correctly, and use [NoContent](https://hackage.haskell.org/package/servant-0.16/docs/Servant-API-ContentTypes.html#t:NoContent) from servant for the response type.

jonathanknowles replied:
Done.

KtorZ commented on src/Cardano/Wallet/Api/Types/Amount.hs:
```diff
@@ -0,0 +1,28 @@
+data Amount = Amount
+    { _quantity :: Natural
+    , _unit :: T.CurrencyUnit
+    } deriving (Eq, Generic, Show)
```
For every types containing a unit, I'll suggest to re-use the approach of the old code base:

-> https://github.com/input-output-hk/cardano-sl/blob/1d2b72f652cbc89691c60a4d6804fc8116d7678c/lib/src/Pos/Util/UnitsOfMeasure.hs

Minus the `Buildable` and `ToSwagger` instances, this code is quite concise and allows us to group the logic of representing types with units.

KtorZ commented on src/Cardano/Wallet/Api/Types/WalletAddressPoolGap.hs:
```diff
@@ -0,0 +1,34 @@
+
+instance ToJSON WalletAddressPoolGap where
+    toJSON = Number
+        . fromIntegral
```
`Number . fromIntegral` could just be `toJSON` here :+1:

jonathanknowles replied:
Nice, thanks!

KtorZ commented on src/Cardano/Wallet/Api/Types/WalletAddressPoolGap.hs:
```diff
@@ -0,0 +1,34 @@
+    ( Generic )
+
+newtype WalletAddressPoolGap = WalletAddressPoolGap
+    { getWalletAddressPoolGap :: AddressPoolGap }
```
You mentioned you weren't satisfied by this. What about something like a newtype `ApiT` or `InApi` to put around types that are used in the Api. That's doesn't change much the implementation in the end but it removes some trouble in finding good names.

jonathanknowles replied:
I think the idea to use `ApiT` (or something similar) is a good one. Though perhaps it rests on the assumption that every API type has a corresponding "normal" type somewhere else? Perhaps this assumption is always valid?

KtorZ commented on src/Cardano/Wallet/Api/Types/WalletAddressPoolGap.hs:
```diff
@@ -0,0 +1,34 @@
+newtype WalletAddressPoolGap = WalletAddressPoolGap
+    { getWalletAddressPoolGap :: AddressPoolGap }
+    deriving stock (Generic, Show)
+    deriving newtype (Bounded, Enum, Eq, Ord)
```
Not sure we need the `Bounded` and `Enum` here :thinking:

jonathanknowles replied:
They are used in `Cardano.Wallet.Api.Arbitraries`:
```hs
instance Arbitrary WalletAddressPoolGap where                                                                                                                                         
    arbitrary = arbitraryBoundedEnum    
```
This particular `Arbitrary` instance is used in JSON round-trip testing.

KtorZ commented on src/Cardano/Wallet/Api/Types/WalletBalance.hs:
```diff
@@ -0,0 +1,24 @@
+
+data WalletBalance = WalletBalance
+    { _available :: !T.Amount
+    , _total :: !T.Amount
```
units are missing from the amounts here :+1:

jonathanknowles replied:
The unit is defined as part of the definition of `Amount`:
```hs
data Amount = Amount
    { _quantity :: Natural
    , _unit :: T.CurrencyUnit
    } deriving (Eq, Generic, Show)

instance FromJSON Amount where
    parseJSON = genericParseJSON defaultRecordTypeOptions
instance ToJSON Amount where
    toJSON = genericToJSON defaultRecordTypeOptions
```
This seems consistent with the Swagger specification. Perhaps I have misunderstood your meaning?

KtorZ commented on src/Cardano/Wallet/Api/JSON.hs:
```diff
@@ -0,0 +1,26 @@
+
+defaultRecordTypeOptions :: Options
+defaultRecordTypeOptions = defaultOptions
+    { fieldLabelModifier = drop 1
```
I guess we also need to transform from `camelCase` to `snake_case` for all fields as well!

jonathanknowles replied:
Done!

KtorZ commented on src/Cardano/Wallet/Api/Types/WalletName.hs:
```diff
@@ -0,0 +1,48 @@
+mkWalletName n
+    | T.length n < walletNameMinLength = Left WalletNameTooShortError
+    | T.length n > walletNameMaxLength = Left WalletNameTooLongError
+    | otherwise = Right $ WalletName n
```
We'll have to move this logic into the wallet primitives. The API shouldn't bother about this and simply delegate the validation of wrapped types.

jonathanknowles replied:
Fair enough! I'll move this type into the `Primitives` module, leaving a thin wrapper in `Api`. (Perhaps another example of where `ApiT a` would be better.)

jonathanknowles committed:
Use `NoContent` for `DeleteWallet` endpoint.

jonathanknowles committed:
Replace `Number . fromIntegral` with `toJSON`.

jonathanknowles committed:
Use `snake_case` instead of `CamelCase` for API record fields.

KtorZ committed:
regroup api types into one module and rename 'Cardano.Wallet.Api' to 'Cardano.Api'

KtorZ committed:
make Cardano.Api compiles after renaming and grouping files

Didn't do in the previous commit not to mess with git and makes
it easy to understand what happened to that file (shows a rename
in the git history).

KtorZ committed:
review WalletState to work with generic instances

KtorZ committed:
review DelegationState to work with generic instances

KtorZ committed:
simplify percentage types to work better with generics

KtorZ committed:
factor types with unit using a polymorphic 'MeasuredIn' type

KtorZ committed:
introduce 'ApiT' and start moving a few types & logic back to the wallet primitives

jonathanknowles committed:
Use formatting consistent with coding standards.

* One space before opening brackets, everywhere.
* No multiple consecutive blank lines.

jonathanknowles committed:
Remove redundant `Generic` constraint.

jonathanknowles committed:
Push units of measure into the JSON instances for each type.

jonathanknowles committed:
Add functions `addUnit` and `removeUnit` to simplify JSON instances for types with units of measure.

jonathanknowles committed:
Improve comments for `MeasuredIn` type and associated functions.

rvl commented on src/Cardano/Api/Types.hs:
```diff
@@ -0,0 +1,283 @@
+instance FromJSON WalletDelegation where
+    parseJSON = genericParseJSON walletDelegationOptions
+instance ToJSON WalletDelegation where
+    toJSON = genericToJSON walletDelegationOptions
```
This is just my opinion, but for small types like this I think it's more straightforward to write instances manually rather than set up the options for `genericToJSON`.

jonathanknowles replied:
I agree, especially in the case where the particular set of options is used just once. Perhaps we can tidy this up a bit in a future PR.

We're going to be adding quite a lot more to the API modules, so there will be lots of opportunities to do some tidying on the way.

rvl reviewed (approved):
Looks pretty good.
Maybe later - some tests for parsing JSON with invalid amounts/percentages/wallet names.

jonathanknowles committed:
Move `Cardano.Api` to `Cardano.Wallet.Api`.

The API we provide is an API for the wallet, rather than for Cardano as a whole.

jonathanknowles committed:
Use `arbitraryBoundedEnum` in instance of `Arbitrary` for `Percentage`.

jonathanknowles committed:
Appease Haskell Stylish.

jonathanknowles committed:
Add `PoolId` data type and use it to simplify `Arbitrary` instance for `WalletDelegation`.

jonathanknowles committed:
Remove unused `Arbitrary` instance for `MeasuredIn` type.

jonathanknowles committed:
Appease Stylish Haskell.
